### PR TITLE
ENT-4988/master Fix process_command return code influencing promise result

### PR DESCRIPTION
**Pull request by nickanderson**
Merge together:
https://github.com/cfengine/enterprise/pull/548

### CODE REVIEW

**Comment by cf-bottom** [comment]
Thanks for submitting a pull request! Maybe @olehermanse can review this?

**Comment by nickanderson** [comment]
@cf-bottom jenkins, please

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2811/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2811/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2811/

**Comment by nickanderson** [comment]
@cf-bottom jenkins, please

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2812/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2812/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2812/

**Comment by olehermanse** [comment]
@cf-bottom jenkins, please.

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2821/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2821/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2821/

**Comment by nickanderson** [comment]
@cf-bottom jenkins, please

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2824/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2824/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2824/

**Comment by nickanderson** [comment]
@cf-bottom jenkins again please

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2826/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2826/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2826/

**Comment by nickanderson** [comment]
@olehermanse i had to port the valgrind test changes over as well.

File: cf-agent/verify_processes.c [hunk 1]
```diff
@@ -231,12 +233,29 @@ static PromiseResult VerifyProcessOp(EvalContext *ctx, const Attributes *a, cons
...
+                         pp->promiser, a->process_stop);
+                    result = PromiseResultUpdate(result, PROMISE_RESULT_FAIL);
+                    return result;
+                  }
                 }
                 else
                 {
                     cfPS(ctx, LOG_LEVEL_ERR, PROMISE_RESULT_FAIL, pp, a,
-                         "Process promise to stop '%s' could not be kept because '%s' the stop operator failed",
+                         "Process promise to stop '%s' could not be kept because '%s' is not executable",
```
Comment (olehermanse) [inline cf-agent/verify_processes.c]: +1

File: cf-agent/verify_processes.c [hunk 2]
```diff
@@ -231,12 +233,29 @@ static PromiseResult VerifyProcessOp(EvalContext *ctx, const Attributes *a, cons
...
-                    ShellCommandReturnsZero(a->process_stop, SHELL_TYPE_NONE);
+                    Log(LOG_LEVEL_DEBUG, "Found process_stop command '%s' is executable.", a->process_stop);
+
+                  if (ShellCommandReturnsZero(a->process_stop, SHELL_TYPE_NONE))
+                  {
+                      cfPS(ctx, LOG_LEVEL_INFO, PROMISE_RESULT_CHANGE, pp, a,
+                           "Promise to stop '%s' repaired, '%s' returned zero",
+                           pp->promiser, a->process_stop);
+                      result = PromiseResultUpdate(result, PROMISE_RESULT_CHANGE);
+                      return result;
```
Comment (olehermanse) [inline cf-agent/verify_processes.c]: You're introducing a memory leak here, by not calling `DeleteItemList()` before returning. Look at the other return statements in the same functions to see what cleanup is necessary.

Also, AFAICT you shouldn't return here, there is more logic later, that would run after the shell command. Most notably related to signals and restarting.

File: cf-agent/verify_processes.c [hunk 3]
```diff
@@ -231,12 +233,29 @@ static PromiseResult VerifyProcessOp(EvalContext *ctx, const Attributes *a, cons
...
+                  }
+                  else
+                  {
+                    cfPS(ctx, LOG_LEVEL_ERR, PROMISE_RESULT_FAIL, pp, a,
+                         "Promise to stop '%s' failed, '%s' returned nonzero",
+                         pp->promiser, a->process_stop);
+                    result = PromiseResultUpdate(result, PROMISE_RESULT_FAIL);
+                    return result;
```
Comment (olehermanse) [inline cf-agent/verify_processes.c]: In this case, it might be appropriate to return early, like you do, since it's erroring. You need to `DeleteItemList(killlist)` first, though.

File: cf-agent/verify_processes.c (hunk 2 above)
Comment (nickanderson) [reply cf-agent/verify_processes.c]: Yeah, I returned from here because of where the else was already returning. I see that there it had DeleteItemList(killlist) though, so Ill add that in.

File: cf-agent/verify_processes.c (hunk 3 above)
Comment (nickanderson) [reply cf-agent/verify_processes.c]: Right, that was my reasoning for the guess.

File: cf-agent/verify_processes.c (hunk 2 above)
Comment (nickanderson) [reply cf-agent/verify_processes.c]: Seems like before  doing the signals (i guess that's what's happening here)

https://github.com/cfengine/core/pull/3855/files#diff-6c1d2f248881f255c48f13ff8f1d5fc6L248

it should only try to signal the processes that still exist, like the process table should be refreshed. It's my current understanding that the process table is only ever refreshed at the start of a bundle. But it seems like if we knowingly alter a process, we should refresh our view of the process state as we do with our package cache.

File: cf-agent/verify_processes.c (hunk 3 above)
Comment (nickanderson) [reply cf-agent/verify_processes.c]: Added

File: cf-agent/verify_processes.c (hunk 2 above)
Comment (olehermanse) [reply cf-agent/verify_processes.c]: I admit, this code is very hard to completely understand, but I think it's always safer to preserve behavior. Clearly, there are not enough tests for this code, but it might be working for people in real policy, and we don't want to break that.

So, I would remove this early return (and the deleteitemlist), at least here, since this is the success path.
Previously it would **not** return early in any of these cases you added. There is more code below for signalling and restarting. With your changes it **always** returns early after the shell command.

Comment (nickanderson) [reply cf-agent/verify_processes.c]: I found in the changelog that 3.3.0 introduced the ability for process\_stop and
signals to co-exist within the same promise.

> -   Allow both \`process\_stop' and \`signals' constraints in
>     \`processes' promises at the same time.

This makes me think that it should *not* return early. It should do both
process\_stop and signals regardless of success or failure in either.

But I guess that it has not been sending signals if process\_stop was not
executable, so I have only removed the early returns that I added.

**Comment by nickanderson** [comment]
@cf-bottom jenkins

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2853/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2853/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2853/

**Comment by nickanderson** [comment]
@cf-bottom jenkins please

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2859/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2859/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2859/

File: cf-agent/verify_processes.c [hunk 4]
```diff
@@ -231,12 +233,29 @@ static PromiseResult VerifyProcessOp(EvalContext *ctx, const Attributes *a, cons
             {
                 if (IsExecutable(CommandArg0(a->process_stop)))
                 {
-                    ShellCommandReturnsZero(a->process_stop, SHELL_TYPE_NONE);
+                    Log(LOG_LEVEL_DEBUG, "Found process_stop command '%s' is executable.", a->process_stop);
+
+                  if (ShellCommandReturnsZero(a->process_stop, SHELL_TYPE_NONE))
```
Comment (olehermanse) [inline cf-agent/verify_processes.c]: Wrong indentation here.

Comment (nickanderson) [reply cf-agent/verify_processes.c]: Finally got cfengine-code-style loaded in emacs, fixed.

**Comment by nickanderson** [comment]
@cf-bottom jenkins one last time while we wait for an approval.

**Comment by nickanderson** [comment]
@olehermanse

- [X] Fixed indentation
- [X] Squashed

Waiting for approval so that I can merge this pr.

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2870/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2870/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2870/

**Comment by nickanderson** [comment]
@cf-bottom Jenkins again?

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2877/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2877/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2877/

File: travis-scripts/valgrind.sh [hunk 5]
```diff
@@ -162,20 +171,24 @@ echo "Running update.cf:"
...
+
 echo "Running cf-check:"
 valgrind $VG_OPTS /var/cfengine/bin/cf-check diagnose /var/cfengine/state/*.lmdb 2>&1 | tee check.txt
 check_output check.txt
 
 echo "Checking that bootstrap ID doesn't change"
-/var/cfengine/bin/cf-agent --show-evaluated-vars | grep bootstrap_id > id_a
-/var/cfengine/bin/cf-agent -K --show-evaluated-vars | grep bootstrap_id > id_b
+/var/cfengine/bin/cf-agent --define mpf_disable_cfe_internal_limit_robot_agents --show-evaluated-vars | grep bootstrap_id > id_a
+/var/cfengine/bin/cf-agent --define mpf_disable_cfe_internal_limit_robot_agents -K --show-evaluated-vars | grep bootstrap_id > id_b
```
Comment (olehermanse) [inline travis-scripts/valgrind.sh]: @nickanderson these changes don't belong in this PR.

**Comment by nickanderson** [comment]
Ok i removed the changes to valgrind.sh and rebased.

**Comment by nickanderson** [comment]
@cf-bottom jenkins

File: tests/acceptance/05_processes/process_stop.cf [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+      # repaired. Note: At the time of authorship there is no validation that
+      # the selected pids were killed by process_stop. We are only using the
+      # return code.
+
+      "."
+        process_stop => "$(G.true)",
+        handle => "process_stop_return_zero_expect_repaired",
+        classes => explicit_results( "namespace", "$(this.handle)_is" );
+
+      # G.false is expected to return true, and the promise is expected to be
```
Comment (vpodzime) [inline tests/acceptance/05_processes/process_stop.cf]: *G.false is expected to return true*

Comment (vpodzime) [reply tests/acceptance/05_processes/process_stop.cf]: The pasta needs some ketching-up.

File: tests/acceptance/05_processes/process_stop.cf [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+                    "process_stop_with_args_return_nonzero_expect_repaired_is_repaired",
+                    "process_stop_return_nonzero_expect_failed_is_failed",
+                    "process_stop_return_zero_expect_repaired_is_repaired",
+                    "process_stop_not_executable_expect_failed_is_failed",
+      };
+
+
+      DEBUG::
+      "found_classes" slist => classesmatching( "process_stop_.*");
+      "difference" slist => difference( found_classes, expected_classes );
```
Comment (vpodzime) [inline tests/acceptance/05_processes/process_stop.cf]: Weird indentation.

File: tests/acceptance/05_processes/process_stop.cf [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+    DEBUG::
+      "Found unexpected class: $(difference)";
+
+    ok::
+      "$(this.promise_filename) Pass";
+    !ok::
+      "$(this.promise_filename) FAIL";
+}
+
+body classes explicit_results(scope, class_prefix)
```
Comment (vpodzime) [inline tests/acceptance/05_processes/process_stop.cf]: Is it possible to share this between tests?

File: tests/acceptance/05_processes/process_stop.cf (hunk 6 above)
Comment (nickanderson) [reply tests/acceptance/05_processes/process_stop.cf]: ack

File: tests/acceptance/05_processes/process_stop.cf (hunk 7 above)
Comment (nickanderson) [reply tests/acceptance/05_processes/process_stop.cf]: ack

File: tests/acceptance/05_processes/process_stop.cf (hunk 8 above)
Comment (nickanderson) [reply tests/acceptance/05_processes/process_stop.cf]: Not from where it currently is. Usually, I use results, which is in the stdlib, but it defined more classes than I wanted in this case, so I made this custom body.

If you want this in dcs lib, then i can do that in a separate pr.
